plotly/graph_reference.py

```diff
@@ -0,0 +1,355 @@
+import json
+import os
+import re
+from pkg_resources import resource_string
+
+import requests
+import six
+
+from plotly import files, utils
+
+GRAPH_REFERENCE_PATH = '/v2/plot-schema'
+
+
+# for backwards compat, we need to add a few class names
+_BACKWARDS_COMPAT_CLASS_NAME_TO_OBJECT_NAME = {
+    'AngularAxis': 'angularaxis',
+    'ColorBar': 'colorbar',
+    'Area': 'scatter',
+    'Histogram2d': 'histogram2d',
+    'Histogram2dContour': 'histogram2dcontour',
+    'RadialAxis': 'radialaxis',
+    'Scatter3d': 'scatter3d',
+    'XAxis': 'xaxis',
+    'XBins': 'xbins',
+    'YAxis': 'yaxis',
+    'YBins': 'ybins',
+    'ZAxis': 'zaxis'
+}
+
+
+def get_graph_reference():
+    """
+    Attempts to load local copy of graph reference or makes GET request if DNE.
+
+    :return: (dict) The graph reference.
+    :raises: (PlotlyError) When graph reference DNE and GET request fails.
+
+    """
+    default_config = files.FILE_CONTENT[files.CONFIG_FILE]
+    if files.check_file_permissions():
+        graph_reference = utils.load_json_dict(files.GRAPH_REFERENCE_FILE)
+        config = utils.load_json_dict(files.CONFIG_FILE)
+
+        # TODO: https://github.com/plotly/python-api/issues/293
+        plotly_api_domain = config.get('plotly_api_domain',
+                                       default_config['plotly_api_domain'])
+    else:
+        graph_reference = {}
+        plotly_api_domain = default_config['plotly_api_domain']
+
+    sha1 = hashlib.sha1(six.b(str(graph_reference))).hexdigest()
+
+    graph_reference_url = '{}{}?sha1={}'.format(plotly_api_domain,
+                                                GRAPH_REFERENCE_PATH, sha1)
+
+    try:
+        response = requests.get(graph_reference_url)
+        response.raise_for_status()
+    except requests.exceptions.RequestException:
+        if not graph_reference:
+            path = os.path.join('graph_reference', 'default-schema.json')
+            s = resource_string('plotly', path).decode('utf-8')
+            graph_reference = json.loads(s)
+    else:
+        if six.PY3:
+            content = str(response.content, encoding='utf-8')
+        else:
+            content = response.content
+        data = json.loads(content)
+        if data['modified']:
+            graph_reference = data['schema']
+
+    return utils.decode_unicode(graph_reference)
+
+
+def object_name_to_class_name(object_name):
+    """
+    Single function to handle turning object names into class names.
+
+    GRAPH_REFERENCE has names like `error_y`, which we'll turn into `ErrorY`.
+
+    :param (str) object_name: Presumably an object_name from GRAPH_REFERENCE.
+    :return: (str)
+
+    """
+    string = _get_hr_name(object_name)
+
+    # capitalize first letter
+    string = re.sub(r'[A-Za-z]', lambda m: m.group().title(), string, count=1)
+
+    # replace `*_<c>` with `*<C>` E.g., `Error_x` --> `ErrorX`
+    string = re.sub(r'_[A-Za-z0-9]+', lambda m: m.group()[1:].title(), string)
+
+    return string
```

Comment:
yeah, exactly. The funky thing about this is *if* we add an hr_name of `x_axis` at some point `XAxis` would go away, making a backwards incompatible change... one way to prevent it is to use the object name *and* the hrName to create classes... which does a *lot* of doubling...

I'm not sure what the right path is here.